rfcs/text/formik.md

```diff
@@ -0,0 +1,695 @@
+const schemaShapeWarning = yup.string().test({
+  test: (shape) => (!['tetrahedron', 'dodecahedron'].includes(shape)),
+  message: 'Choosing a ${value} is not recommended! But form submission is allowed.'
+});
+
+const schemaShapeInfo = yup.string().test({
+  test: (shape) => (!['octahedron', 'icosahedron'].includes(shape)),
+  message: "You've chosen an ${value}. That's a nice shape. Form submission is allowed."
+});
+
+const App = () => {
+
+  const [shapeInfo,    setShapeInfo]    = React.useState(null);
+  const [shapeWarning, setShapeWarning] = React.useState(null);
+
+  /** @see https://jaredpalmer.com/formik/docs/api/field#validate */
+  const validateShape = (shape) => {
+
+    setShapeInfo(null);
+    setShapeWarning(null);
+
+    // Error (blocking) validation:
+    if (['cube', 'tesseract'].includes(shape)) {
+      return `You cannot choose a ${shape}! Form submission is blocked!`;
+    }
+
+    // Warning (non-blocking) validation:
+    try {
+      /** @see https://github.com/jquense/yup/blob/v0.27.0/README.md#mixedvalidatesyncvalue-any-options-object-any */
+      schemaShapeWarning.validateSync(shape);
+    } catch (err) {
+      setShapeWarning(err.message);
+    }
+
+    // Info (non-blocking) validation:
+    try {
+      schemaShapeInfo.validateSync(shape);
+    } catch (err) {
+      setShapeInfo(err.message);
+    }
+
+  };
+
+  return (
+    <Formik
+      initialValues={{ shape: 'sphere' }}
+      onSubmit={() => window.alert('Submitting')}
+    >
+      {({ errors }) => (
+        <Form>
+          <Field as="select" name="shape" validate={validateShape}>
+            {shapes.map((shape) => (<option value={shape} key={shape}>{shape}</option>))}
+          </Field>
+          <div style={{ color: 'red'    }}>{errors.shape}</div>
+          <div style={{ color: 'orange' }}>{shapeWarning}</div>
+          <div style={{ color: 'blue'   }}>{shapeInfo}</div>
+          <input type="submit" />
+        </Form>
+      )}
+    </Formik>
+  );
+
+};
+
+render(<App />, document.getElementById('root'));
+```
+
+
+# Alternatives
+
+There are a number of alternative packages available to developers looking for a solution to web forms and validation. It is our intention to provide an API that affords developers the flexibility to choose whatever solution they want to and integrate their choice with Carbon's Form and input components. 
+
+The alternative to integrating Formik is to continue to maintain the current implementation of Form and Validations in Carbon. We would then continue to add additional functionality to meet any new requirements that may arise. This currently already adds a significant overhead to the team any will only grow as new functionality is implemented. A further issue is that it will continue to cause vendor lock-in, forcing consumers to use what is provided by Carbon out-of-the-box, and may hurt its adoption as a result.
+
+
+# Adoption strategy
```

Comment:
We will need to think about what this looks like to migrate from how forms are built today - to how forms are built tomorrow using Formik. I.e. if i'm a developer with a project using v12.0.0 of Carbon, what do i need to do to upgrade to the version with Formik support.